client/container_list.go

```diff
@@ -35,7 +37,14 @@ func (cli *Client) ContainerList(ctx context.Context, options types.ContainerLis
 	}
 
 	if options.Filter.Len() > 0 {
-		filterJSON, err := filters.ToParam(options.Filter)
+		filterJSON := ""
+		err := errors.New("")
+		// for daemons older than v1.10, filter must be of the form map[string][]string
+		if cli.version != "" && version.Version(cli.version).LessThan(version.Version("1.10")) {
```

Comment:
Sorry I think I wasn't very clear before.

For each engine in the cluster, there indeed is a dedicated apiClient for that engine, but with two extra details:
1. We only find out the engine version after the client has been created
2. Since engines in the cluster can go down and come back up after being updated, we'd like to be able to update the API version if needed.

Beyond these two cases, things are fine, and work exactly as you pointed out.